poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+ */
+func (p *Promise) UnmarshalInit(shareSuite, keySuite abstract.Suite) *Promise {
+	p.shareSuite = shareSuite
+	p.keySuite = keySuite
+	return p
+}
+
+/* Verifies that the Promise was constructed correctly.
+ *
+ * Arguments
+ *    promiserKey = the long term key the caller believes the Promise to be from
+ *
+ * Return
+ *   an error if the promise is malformed, nil otherwise.
+ *
+ * TODO Consider more ways to verify (such as making sure there are no duplicate
+ *      keys in p.insurers or that the promiser's long term public key is not in
+ *      p.insurers).
+ */
+func (p *Promise) VerifyPromise(promiserKey abstract.Point) error {
+	// Verify t <= r <= n
+	if p.t > p.n || p.t > p.r || p.r > p.n {
+		return errors.New("Invalid t-of-n shares Promise. Expected: t <= r <= n")
+	}
+	if !promiserKey.Equal(p.pubKey) {
+		return errors.New("Long term public key of Promise differs from what is expected.")
+	}
+	// There should be a secret and public key for each of the n insurers.
+	if len(p.insurers) != p.n || len(p.secrets) != p.n {
+		return errors.New("Insurers and secrets array should be of length promise.n")
+	}
+	return nil
+}
+
+// Returns the id of the Promise
+func (p *Promise) GetId() string {
+	return p.id
+}
+
+/* Given a Diffie-Hellman shared secret, encrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to encrypt
+ *    diffieBase  = the DH shared secret
+ *
+ * Return
+ *   the encrypted secret
+ */
+func (p *Promise) diffieHellmanEncrypt(secret abstract.Secret,
+	diffieBase abstract.Point) abstract.Secret {
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman given.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Add(secret, diffieSecret)
+}
+
+/* Given a Diffie-Hellman shared secret, decrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to decrypt
+ *    diffieBase  = the DH shared secret
+ *
+ * Return
+ *   the decrypted secret
+ */
+func (p *Promise) diffieHellmanDecrypt(secret abstract.Secret,
+	diffieBase abstract.Point) abstract.Secret {
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman given.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Sub(secret, diffieSecret)
+}
+
+/* Verifies that a share has been properly constructed. This should be called by
+ * insurers to verify that the share they received is valid constructed.
+ *
+ * Arguments
+ *    i         = the index of the share to verify
+ *    gKeyPair  = the long term key pair of the insurer of share i
+ *
+ * Return
+ *  an error if the share is malformed, nil otherwise.
+ *
+ * Note
+ *   Make sure that the proper index and key is specified.
+ */
+func (p *Promise) VerifyShare(i int, gKeyPair *config.KeyPair) error {
```

Comment:
It would be more difficult to make this private.

VerifyShare needs to know the index and keypair of the insurer who was unmarshalling. I would need to store this information within the Promise since I couldn't alter the marshalling interface to include them. However, I thought that these two would be better suited as arguments versus struct variables.
 
Different actions will be taken depending upon whether or not this function succeeds. If it fails, then the share and/or Promise was improperly constructed (likely maliciously). Hence, the insurer could construct a BlameProof to show that the Promise is untrustworthy at best or malicious at worst.
It would be more difficult to make this private.

VerifyShare needs to know the index and keypair of the insurer who was unmarshalling. I would need to store this information within the Promise since I couldn't alter the marshalling interface to include them. However, I thought that these two would be better suited as arguments versus struct variables.
 
Different actions will be taken depending upon whether or not this function succeeds. If it fails, then the share and/or Promise was improperly constructed. Hence, the insurer could construct a BlameProof to show that the Promise is untrustworthy at best or malicious at worst. If it succeeds, then the Promise is good and a PromiseSignature can be constructed.